File: razarion-client/src/main/java/com/btxtech/client/cockpit/radar/AbstractGameCoordinates.java
package com.btxtech.client.cockpit.radar;

import com.btxtech.shared.datatypes.DecimalPosition;
import com.btxtech.shared.datatypes.Rectangle2D;
import com.btxtech.uiservice.control.GameUiControl;
import elemental.html.CanvasRenderingContext2D;

import javax.inject.Inject;

/**
 * Created by Beat
 * on 20.08.2017.
 */
public abstract class AbstractGameCoordinates extends AbstractMiniMap {
    // private Logger logger = Logger.getLogger(AbstractGameCoordinates.class.getName());
    @Inject
    private GameUiControl gameUiControl;

    @Override
    protected void setupTransformation(double zoom, CanvasRenderingContext2D ctx, int width, int height) {
        DecimalPosition planetSize = gameUiControl.getPlanetConfig().getSize();

        double scale = setupGameScale();
        DecimalPosition centerOffset = getViewField().calculateCenter();

        float xShift = setupXShift(width, planetSize.getX(), scale, centerOffset);
        float yShift = setupYShift(height, planetSize.getY(), scale, centerOffset);

        ctx.scale((float) scale, (float) -scale);
        ctx.translate(-xShift, -yShift);
    }

    public DecimalPosition canvasToReal(DecimalPosition canvasPosition) {
        DecimalPosition planetSize = gameUiControl.getPlanetConfig().getSize();

        double scale = setupGameScale();
        DecimalPosition real = canvasPosition.divide(scale, -scale);
        DecimalPosition centerOffset = getViewField().calculateCenter();
        real = real.add(setupXShift(getWidth(), planetSize.getX(), scale, centerOffset), setupYShift(getHeight(), planetSize.getY(), scale, centerOffset));
        return real;
    }

    protected Rectangle2D getVisibleField() {
        return new Rectangle2D(canvasToReal(new DecimalPosition(0, getHeight())), canvasToReal(new DecimalPosition(getWidth(), 0)));
    }

    protected float toCanvasPixel(double pixels) {
        return (float) (pixels / setupGameScale());
    }

    private float setupXShift(int width, double playWidth, double scale, DecimalPosition centerOffset) {
        float xDownerLimit = (float) (width / scale / 2.0);
        float xUpperLimit = (float) (playWidth - xDownerLimit);
        float xShift;
        if (centerOffset.getX() < xDownerLimit) {
            xShift = 0.0f;
        } else if (centerOffset.getX() > xUpperLimit) {
            xShift = xUpperLimit - xDownerLimit;
        } else {
            xShift = (float) (centerOffset.getX() - xDownerLimit);
        }
        return xShift;
    }

    private float setupYShift(int height, double playHeight, double scale, DecimalPosition centerOffset) {
        float yDownerLimit = (float) (height / scale / 2.0);
        float yUpperLimit = (float) (playHeight - yDownerLimit);
        float yShift;
        if (centerOffset.getY() < yDownerLimit) {
            yShift = (float) (playHeight - yUpperLimit + yDownerLimit);
        } else if (centerOffset.getY() > yUpperLimit) {
            yShift = (float) playHeight;
        } else {
            yShift = (float) (centerOffset.getY() + yDownerLimit);
        }
        return yShift;
    }
}

File: razarion-frontend/src/app/game/cockpit/main/radar/abstract-game-coordinates.ts
import { DecimalPosition, GameUiControl } from 'src/app/gwtangular/GwtAngularFacade';
import { AbstractMiniMap } from './abstract-mini-map';
import { BabylonRenderServiceAccessImpl } from 'src/app/game/renderer/babylon-render-service-access-impl.service';
import { GwtInstance } from 'src/app/gwtangular/GwtInstance';

export abstract class AbstractGameCoordinates extends AbstractMiniMap {
    constructor(gameUiControl: GameUiControl, renderService: BabylonRenderServiceAccessImpl) {
        super(gameUiControl, renderService);
    }

    protected setupTransformation(zoom: number, ctx: CanvasRenderingContext2D, width: number, height: number): void {
        let planetSize: DecimalPosition = this.gameUiControl.getPlanetConfig().getSize();

        let scale: number = this.setupGameScale();

        let xShift: number = this.setupXShift(width, planetSize.getX(), scale, this.getViewField().getCenter());
        let yShift: number = this.setupYShift(height, planetSize.getY(), scale, this.getViewField().getCenter());


        ctx.scale(scale, -scale);
        ctx.translate(-xShift, -yShift);
    }

    public canvasToReal(canvasPositionX: number, canvasPositionY: number): DecimalPosition {
        let planetSize: DecimalPosition = this.gameUiControl.getPlanetConfig().getSize();

        let scale: number = this.setupGameScale();
        let real = GwtInstance.newDecimalPosition(canvasPositionX / scale, canvasPositionY / -scale);
        real = real.add(this.setupXShift(this.getWidth(), planetSize.getX(), scale, this.getViewField().getCenter()), this.setupYShift(this.getHeight(), planetSize.getY(), scale, this.getViewField().getCenter()));
        return real;
    }

    protected toCanvasPixel(pixels: number): number {
        return pixels / this.setupGameScale();
    }

    private setupXShift(width: number, planetSizeX: number, scale: number, centerOffset: DecimalPosition): number {
        let xDownerLimit: number = width / scale / 2.0;
        let xUpperLimit: number = planetSizeX - xDownerLimit;
        let xShift: number;
        if (centerOffset.getX() < xDownerLimit) {
            xShift = 0.0;
        } else if (centerOffset.getX() > xUpperLimit) {
            xShift = xUpperLimit - xDownerLimit;
        } else {
            xShift = centerOffset.getX() - xDownerLimit;
        }
        return xShift;
    }

    private setupYShift(height: number, playHeight: number, scale: number, centerOffset: DecimalPosition): number {
        let yDownerLimit: number = height / scale / 2.0;
        let yUpperLimit: number = playHeight - yDownerLimit;
        let yShift: number;
        if (centerOffset.getY() < yDownerLimit) {
            yShift = playHeight - yUpperLimit + yDownerLimit;
        } else if (centerOffset.getY() > yUpperLimit) {
            yShift = playHeight;
        } else {
            yShift = centerOffset.getY() + yDownerLimit;
        }
        return yShift;
    }
}